File: fulltank/fulltank.py
#!/usr/bin/env python

from collections import namedtuple, defaultdict
from heapq import *
import sys


def print_graph(G):
    """Print the graph to standard out."""
    for v in G.keys():
        for e in G[v].keys():
            print("%4d - (%3d) -> %4d" % (v, G[v][e], e))
        if not G[v]:
            print("%4d - (X)" % v)


def shortest_path(G, prices, src, dst, cap):
    """Use Dijkstra's algorithm to compute the cheapest way to the destination.

    .. Keyword Arguments:
    :param: G: The graph to traverse.
    :param: prices: The fuelling prices at each node.
    :param: src: The source node to start from.
    :param: dst: The destination.
    :param: cap: The capacity of the traveller.

    .. Types:
    :type: G: A dictionary of integers to a dictionary of integers.
    :type: prices: A dictionary mapping of integer to integers.
    :type: src: An integer.
    :type: dst: An integer.
    :type: cap: An integer.

    .. Returns:
    :returns: A tuple with a list of costs and list of previous nodes.

    """
    Vertex  = namedtuple("Vertex", ["cost", "id", "tank"])
    visited = defaultdict(lambda: False)
    cost    = defaultdict(lambda: float("inf"))

    Q = []
    cost[src, 0] = 0

    heappush(Q, Vertex(0, src, 0,))
    while Q:

        u = heappop(Q)

        if visited[u.id, u.tank]:
            continue

        if u.id == dst:
            break

        # Check if we can move to a neighboring vertex.
        if (u.tank < cap and
            u.cost + prices[u.id] < cost[u.id, u.tank+1]):
            new_cost = u.cost + prices[u.id]
            cost[u.id, u.tank+1] = new_cost
            heappush(Q, Vertex(new_cost, u.id, u.tank+1))

        for v,d in G[u.id].items():
            if (u.tank >= d and
                cost[u.id, u.tank] < cost[v, u.tank - d]):
                cost[v, u.tank - d] = cost[u.id, u.tank]
                heappush(Q, Vertex(cost[v, u.tank - d], v, u.tank - d))

        visited[u.id,u.tank] = True
    return cost[dst, 0]


def main():
    """Create a graph over several cities and compute the optimal fuel-path.

    """
    # Read the initial graph.
    n, m = (int(i) for i in sys.stdin.readline().split())
    prices = { i : int(p) for i,p in enumerate(sys.stdin.readline().split()) }
    G = { v : dict() for v in range(0, len(prices)) }
    for _ in range(0, m):
        u, v, d = (int(i) for i in sys.stdin.readline().split())
        G[u][v] = d
        G[v][u] = d

    q = int(sys.stdin.readline())
    for _ in range(0, q):
        c, s, e = (int(i) for i in sys.stdin.readline().split())
        cost = shortest_path(G, prices, s, e, c)
        if cost < float("inf"):
            print(cost)
        else:
            print("impossible")

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: fulltank/fulltank.cpp
#include <functional>
#include <iostream>
#include <vector>
#include <queue>
#include <map>

using namespace std;

const unsigned infinity = 1 << 28;


/** @brief Print the graph to standard out.

    @param G The graph to be printed.
*/
void print_graph(vector<vector<pair<unsigned, unsigned> > > &G)
{
    for (size_t i = 0; i < G.size(); ++i)
    {
        for (size_t j = 0; j < G[i].size(); ++j)
        {
            pair<unsigned, unsigned> p = G[i][j];
            cout << i << " - " << p.second << " -> " << p.first << endl;
        }
        if (G[i].empty())
        {
            cout << i << " - (X)" << endl;
        }
    }
}


struct Vertex
{
    Vertex(unsigned id, unsigned tank, unsigned cost) :
        id(id), tank(tank), cost(cost) {}
    unsigned id;
    unsigned tank;
    unsigned cost;
};

bool operator>(const Vertex &v0, const Vertex &v1)
{
    return v0.cost > v1.cost;
}


/** @brief Use Dijkstra's algorithm to compute the cheapest way to dst.

    @param G      The graph with the roads and lengths.
    @param prices The prices to fuel at the i:th city.
    @param src    The source city to start from
    @param dst    The destination city to look for.
    @param cap    The capacity of the vehicle.
    @return The cost to reach the destination, or 'infinity' if unreachable.
*/
unsigned shortest_path(vector<vector<pair<unsigned, unsigned> > > &G,
                       vector<unsigned> prices,
                       unsigned src,
                       unsigned dst,
                       unsigned cap)
{
    vector<vector<unsigned> > cost(G.size(), vector<unsigned>(100, infinity));
    vector<vector<bool> >     visited(G.size(), vector<bool>(100, false));
    priority_queue<Vertex, vector<Vertex>, greater<Vertex> > Q;

    cost[src][0] = 0;
    Q.push(Vertex(src, 0, 0));

    while (!Q.empty())
    {
        Vertex u = Q.top();
        Q.pop();
        // cout << u.id << " " << u.tank << " " << u.cost << endl;

        if (visited[u.id][u.tank])
        {
            continue;
        }

        if (dst == u.id)
        {
            break;
        }

        // Check if we can move to a neighboring vertex.
        if (u.tank < cap &&
            u.cost + prices[u.id] < cost[u.id][u.tank + 1])
        {
            unsigned new_cost = u.cost + prices[u.id];
            cost[u.id][u.tank + 1] = new_cost;
            Q.push(Vertex(u.id, u.tank + 1, new_cost));
        }

        for (size_t i = 0; i < G[u.id].size(); ++i)
        {
            unsigned v = G[u.id][i].first;
            unsigned d = G[u.id][i].second;
            if (u.tank >= d &&
                cost[u.id][u.tank] < cost[v][u.tank - d])
            {
                cost[v][u.tank - d] = cost[u.id][u.tank];
                Q.push(Vertex(v, u.tank - d, cost[v][u.tank - d]));
            }
        }
        visited[u.id][u.tank] = true;
    }
    return cost[dst][0];
}


int main()
{
    unsigned n = 0;
    unsigned m = 0;
    cin >> n >> m;

    vector<unsigned> prices(n);
    for (size_t i = 0; i < n; ++i)
    {
        cin >> prices[i];
    }

    vector<vector<pair<unsigned, unsigned> > > G(n);

    for (size_t i = 0; i < m; ++i)
    {
        unsigned u = 0;
        unsigned v = 0;
        unsigned d = 0;
        cin >> u >> v >> d;
        G[u].push_back(make_pair(v, d));
        G[v].push_back(make_pair(u, d));
    }

    unsigned q = 0;
    cin >> q;
    for (size_t i = 0; i < q; ++i)
    {
        unsigned c = 0;
        unsigned s = 0;
        unsigned e = 0;
        cin >> c >> s >> e;
        unsigned cost = shortest_path(G, prices, s, e, c);
        if (cost < infinity)
        {
            cout << cost << endl;
        }
        else
        {
            cout << "impossible" << endl;
        }
    }

    return 0;
}
